src/components/custom-field-input-date/custom-field-input-date.jsx

```diff
@@ -0,0 +1,142 @@
+import React, { useEffect, useRef, useState } from 'react';
+import PropTypes from 'prop-types';
+import calendarSvg from '../../svgs/icon-calendar-fill.svg';
+import Icon from '../icon/icon.jsx';
+import CustomFieldInputText from '../custom-field-input-text/custom-field-input-text.jsx';
+import { convertToFormat, validDate } from './format/format-date.js';
+
+const isValidInput = (value) => {
+  if (value === '' || value === undefined) {
+    return true;
+  }
+
+  return validDate(value);
+};
+
+const isValueValid = (value, error, isInputValid = false) => {
+  if (error) {
+    return false;
+  }
+
+  if (!isInputValid) {
+    return false;
+  }
+
+  return isValidInput(value);
+};
+
+export default function CustomFieldInputDate(props) {
+  const inputRef = useRef(null);
+  const [isEditing, setIsEditing] = useState(true);
+  const [isValid, setIsValid] = useState(isValueValid(props.value, props.error, true));
+  const [isFocused, setIsFocused] = useState(false);
+
+  useEffect(() => {
+    const isInputValid = inputRef.current.validity.valid;
+    const valid = isValueValid(props.value, props.error, isInputValid);
+    setIsValid(valid);
+
+    if (!valid) {
+      setIsEditing(true);
+    } else {
+      setIsEditing(false);
+    }
+  }, [inputRef.current]);
+
+  useEffect(() => {
+    if (isEditing && isFocused) {
+      inputRef.current.focus();
+    }
+  }, [isEditing, isFocused]);
+
+  const handleOnFocus = () => {
+    setIsFocused(true);
+    setIsEditing(true);
+  };
+
+  const handleOnBlur = () => {
+    setIsFocused(false);
+    setIsEditing(false);
+  };
+
+  const handleOnChange = (event) => {
+    if (inputRef && inputRef.current) {
+      const isInputValid = inputRef.current.validity.valid;
+      const newDate = convertToFormat(event.target.value, 'yyyy-mm-dd');
+      setIsValid(isValueValid(newDate, props.error, isInputValid) && validDate(newDate));
+    }
+
+    props.onChange(event);
+  };
+
+  const helpText = () => {
+    if (!isValid && !isValidInput(props.value)) {
+      return `"${props.value}" is an invalid date`;
+    }
+
+    return props.helpText;
+  };
+
+  const sharedProps = {
+    className: props.className,
+    disabled: true,
+    icon: <Icon name={calendarSvg.id} title={props.label} stroke="primary" />,
+    label: props.label,
+    inputRef,
+    required: props.required,
+  };
+
+  if (isEditing || !isValid) {
+    const value = validDate(props.value) ? convertToFormat(props.value, 'yyyy-mm-dd') : props.value;
+
+    return (<CustomFieldInputText
+      {...sharedProps}
+      error={!isValid}
+      helpText={helpText()}
+      id={props.id}
+      key={`${props.id}-editing`}
+      min={convertToFormat(props.min, 'yyyy-mm-dd')}
+      max={convertToFormat(props.max, 'yyyy-mm-dd')}
+      onBlur={handleOnBlur}
+      onChange={e => handleOnChange(e)}
```

Comment:
As far as I know, we shouldn't be creating inline anonymous functions for event handlers. Since the `handleOnChange` function already implies handling an event, its signature should just match the event handling. 

```suggestion
      onChange={handleOnChange}
```